MSC4183: Additional Error Codes for submitToken endpoint

dbkr opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-spec-proposals/blob/dbkr/threepid_submittoken_errors/proposals/4183-submitToken-error-codes.md)

# Impls:
Sydent: https://github.com/matrix-org/sydent/pull/592
Synapse: https://github.com/element-hq/synapse/pull/17625
Element Web/Desktop: https://github.com/matrix-org/matrix-react-sdk/pull/12936

----

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/4183#issuecomment-2315935635)

dbkr committed:
MSC4180: Additional Error Codes for submitToken endpoint

dbkr committed:
Bah, markdown

dbkr committed:
Line wrapping

dbkr committed:
Typos

turt2live commented:
I don't feel as though implementation is required for this MSC.

@mscbot fcp merge

Johennes commented on proposals/4180-submitToken-error-codes.md:
```diff
@@ -0,0 +1,52 @@
+# MSC4180: Additional Error Codes for submitToken endpoint
```
```suggestion
# MSC4183: Additional Error Codes for submitToken endpoint
```

I think it's 4183?

dbkr replied:
thanks, must have been some issues filed this time!

dbkr committed:
Right MSC number

Co-authored-by: Johannes Marbach <n0-0ne+github@mailbox.org>

dbkr committed:
Rename

dbkr committed:
Add http status code

anoadragon453 commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,56 @@
+## Unstable prefix
+
+No unstable prefix is deemed necessary. Sydent already sends the common error codes and also sends
+`M_NO_VALID_SESSION` if the code is incorrect. Once an identity server (or homeserver) switches to
+use the new error code, they may not recognise the error condition correctly until updated to support
+the new code. We say that this is acceptable in favour of avoiding the complexity of negotiating error
```
> Once an identity server (or homeserver) switches to use the new error code, they may not recognise the error condition correctly until updated to support the new code.

I'm failing to understand what this means. Is "they" the homeserver?

dbkr replied:
Yeah, that wasn't super clear. Is that better?

anoadragon453 commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,56 @@
+
+## Potential issues
+
+None forseen.
+
+## Alternatives
+
+None considered.
+
+## Security considerations
+
+None forseen.
```
```suggestion
None foreseen.

## Alternatives

None considered.

## Security considerations

None foreseen.
```

anoadragon453 commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,56 @@
+## Proposal
+
+Add the following specific error code as a code that can be returned by the two endpoints given above:
+ * `M_TOKEN_INCORRECT`: Indicates that the token that the user entered to validate the session is incorrect.
```
Would it be worth re-using/redefining `M_UNKNOWN_TOKEN` from the C-S API?

dbkr replied:
Oh, yes, probably

dbkr replied:
On second thoughts... that's talking about an access token whereas this is the user entering a code that's been sent to them, so they're not really the same thing. I think maybe it's clearer if it's a separate code? I can update the MSC to call this out if you agree.

dbkr replied:
Noted this in c026da70

anoadragon453 commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,56 @@
+use the new error code, they may not recognise the error condition correctly until updated to support
+the new code. We say that this is acceptable in favour of avoiding the complexity of negotiating error
+codes with API versions. Since the identity server is generally used via the homeserver now, most
+uses of this API will not currently be receiving a sensible error code in this situation anyway.
```
```suggestion
users of this API will not currently receive a sensible error code in this situation anyway.
```

dbkr committed:
Apply suggestions from code review

Spelling / grammar fixes

Co-authored-by: Andrew Morgan <1342360+anoadragon453@users.noreply.github.com>

dbkr committed:
Word wrap

dbkr committed:
Clearer wording

dbkr committed:
clarify

dbkr commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,56 @@
+although that MSC is for `requestToken` on the C/S API only.
+
+The [`POST /_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken) endpoint in the C/S API also specifies a `submit_url` response parameter, defining its parameters to
+be the same as the Identity API's `submitToken` endpoints. This MSC also affects this.
```
Done

dbkr committed:
Clarify

dbkr commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,56 @@
+HTTP status code 400 should also be used for both of these errors.
+
+Also change the C/S API's definition of [`POST /_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken)
+to specify that the entire API is the same, including response / error codes, rather than just parameters.
```
Hopefully clearer?

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,62 @@
+
+HTTP status code 400 should also be used for both of these errors.
+
+Also change the C/S API's definition of [`POST
+/_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken)
+to specify that the endpoint is the same as the I/S API version in all ways, including response / error codes, rather than just parameters.
```
wait, what? this came out of left field.  Why are we suddenly changing `requestToken` in an MSC which purports to be about `submitToken`? 

There doesn't seem to be any justification for this in the proposal, or discussion of it in general.

What exactly would need to change to bring the current definition into line with the I/S API?

richvdh replied:
ohhh by "the endpoint" do you mean "the endpoint referenced by the `submit_url` response field"? Could you say so if so?

Still, presumably you don't mean "in all ways" -- for example, it doesn't need an IS access token. Would be good to tighten this up a bit.

dbkr replied:
Yep, indeed, that wasn't very clear, hopefully better now.

richvdh commented:
@mscbot concern lack of clarity about changes to requestToken

dbkr committed:
Add note about not reusing error code

dbkr committed:
Clarify that we're taking about the submittoken api in requestToken

dbkr committed:
spelling

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,65 @@
+
+HTTP status code 400 should also be used for both of these errors.
+
+Also change the corresponding submitYoken-like API given in the C/S API's definition of [`POST
+/_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken)
+to specify that both request and response parameters / error codes are the same as the I/S API version, rather than just parameters, apart from needing an I/S API token.
```
```suggestion
Also change the `submit_url` field in the response to [`POST
/_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken),
to specify that response parameters and error codes are the same as the I/S API version, as well as request parameters.
```

richvdh replied:
Are you sure about this, by the way? AFAICT synapse currently returns a bit of HTML here.

richvdh replied:
I don't think the sample impl (https://github.com/element-hq/synapse/pull/17625) actually changes the email submit-token endpoint.

dbkr replied:
So the deal here is that the POST endpoint isn't proxied by Synapse: it doesn't need to be as Sydent's email can just link straight to its own version of the GET endpoint. I've tried to clarify this in 63dcbe09.

dbkr committed:
Clearer wording

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

dbkr committed:
Add note on POST email submitToken api

and how it isn't really a thing in practice

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,70 @@
+endpoint in the C/S API also specifies a `submit_url` response parameter, defining its parameters to be the same as the
+Identity API's `submitToken` endpoints. Everything this MSC specifies applies to this endpoint in the same way.
+
+Note that the email `submitToken` endpoint ([`POST
```
it's a bit confusing that we jump from talking about an endpoint in the C-S API to one in the IS API without mentioning that.

Also, if you're talking specifically about the POST variant, you better say that, rather than saying that the entire endpoint is unused.

dbkr replied:
Yes - sort of, I don't think there is a POST variant for email on the C/S API. Hopefully clearer in af1dc860.

richvdh replied:
This is still confusing: we still jump from talking about an endpoint in the C-S API to one in the IS API without mentioning that.

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,70 @@
+
+HTTP status code 400 should also be used for both of these errors.
+
+Also change the `submit_url` field in the response to [`POST
+/_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken),
+to specify that response parameters and error codes are the same as the I/S API version, as well as request parameters.
```
why does this only apply to `POST /_matrix/client/v3/account/3pid/email/requestToken`, and not also:

* `POST /_matrix/client/v3/account/password/email/requestToken`
* `POST /_matrix/client/v3/account/password/msisdn/requestToken`
* `POST /_matrix/client/v3/register/email/requestToken`
* `POST /_matrix/client/v3/register/msisdn/requestToken`
* `POST /_matrix/client/v3/account/3pid/msisdn/requestToken`

?

dbkr replied:
Mmm, yeah - I'd done it as only editing what I needed to, but you're right, they should be consistent. Changed in 0f43a510.

dbkr committed:
Hopefully make more clearer

dbkr committed:
Apply same change to other submitToken endpoints

...from the other requestToken enpoints

Also try to further clarify the note about the largely unused POST email submitToken

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,79 @@
+obviously, that the token the user entered was incorrect.
+
+This MSC can be considered similar to [MSC4178](https://github.com/matrix-org/matrix-spec-proposals/pull/4178) although
+that MSC is for `requestToken` on the C-S API only.
+
+The [`POST
+/_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken)
+The numerous `requestToken` endpoints (enumerated in the proposal section) in the C/S API also specify a `submit_url`
```
this makes no sense

richvdh replied:
nit: it would be nice if we could decide whether it's C-S API or C/S API, rather than using both in adjacent paragraphs

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,79 @@
+
+## Proposal
+
+Add the following specific error code as a code that can be returned by the two endpoints given above:
```
not least because there are now at least three different endpoints given above, I think you should restate which endpoints are changing here.

dbkr replied:
fixed

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,79 @@
+
+Also apply the same change to all the `submit_url` fields in the various response to the `POST requestToken` endpoints, ie:
+
+ * [`POST /_matrix/client/v3/register/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3registeremailrequesttoken)
+ * [`POST /_matrix/client/v3/register/msisdn/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3registerrequesttoken)
+ * [`POST /_matrix/client/v3/account/3pid/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidemailrequesttoken)
+ * [`POST /_matrix/client/v3/account/3pid/msisdn/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3account3pidmsisdnrequesttoken)
+ * [`POST /_matrix/client/v3/account/password/email/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3accountpasswordemailrequesttoken)
+ * [`POST /_matrix/client/v3/account/password/msisdn/requestToken`](https://spec.matrix.org/v1.11/client-server-api/#post_matrixclientv3accountpasswordmsisdnrequesttoken)
```
haven't checked this morning, but I'm pretty sure your synapse implementation PR doesn't change all of these.

(I don't think synapse returns a `submit_url` from the `email/requestToken` endpoints, but it does from the `msisdn/requestToken` endpoints, and you haven't updated all the implementations of those submit_url endpoints.)

dbkr replied:
It actually does (so it's just as well we included the other endpoints): the various different rest classes use that function in the handler to do the request to the ID server and then the error code is in the exception that it throws.

richvdh replied:
Right, so it does. To be clear:

* Synapse does not return a `submit_url` from any of the `POST /_matrix/client/v3*/email/requestToken` endpoints.
* Synapse does not support `POST /_matrix/client/v3/account/3pid/msisdn/requestToken` (removed in https://github.com/matrix-org/synapse/pull/6101)
* Both `POST /_matrix/client/v3/register/msisdn/requestToken` and `POST /_matrix/client/v3/account/3pid/msisdn/requestToken` use `IdentityHandler.requestMsisdnToken`.

  `requestMsisdnToken` sets `submit_url` to `/_matrix/client/unstable/add_threepid/msisdn/submit_token` (yes, the register endpoint uses the `add_threepid` callback).

  And `/_matrix/client/unstable/add_threepid/msisdn/submit_token` is updated by https://github.com/element-hq/synapse/pull/17625.

richvdh commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,79 @@
+
+HTTP status code 400 should also be used for both of these errors.
+
+Also apply the same change to all the `submit_url` fields in the various response to the `POST requestToken` endpoints, ie:
```
seems like you're only talking about the `POST requestToken` endpoints *in the C-S API*

```suggestion
Also apply the same change to all the `submit_url` fields in the various response to the `POST requestToken` endpoints in the client-server API, i.e.:
```

dbkr committed:
Remove stray lines, more clarifications & consistency fix

dbkr committed:
clarify only c/s api

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

turt2live reviewed (approved):
re-asserting approval as a read marker

richvdh reviewed (approved):
I am now finally up to speed, and other than a few minor editorial nits, this lgtm

richvdh commented:
@mscbot resolve lack of clarity about changes to requestToken

dbkr committed:
Typos & clarity

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

dbkr committed:
Clarify

dbkr commented on proposals/4183-submitToken-error-codes.md:
```diff
@@ -0,0 +1,81 @@
+/_matrix/identity/v2/validate/email/submitToken`](https://spec.matrix.org/v1.11/identity-service-api/#post_matrixidentityv2validateemailsubmittoken))
+is not generally used in practice: Sydent's emails includes a link to click instead of the `submit_url` response field and
+therefore use the `GET` version. Synapse does not implement the `POST` API for email validation for this reason. This
+proposal updates both for consistency.
```
Hopefully clearer

zecakeh commented:
Spec PR: https://github.com/matrix-org/matrix-spec/pull/2277

NaixWintage-art reviewed (commented):
Widebang dose humain
